torchmultimodal/modules/layers/transformer_decoder.py

```diff
@@ -0,0 +1,199 @@
+    Reference: `"Gaussian error linear units"<https://arxiv.org/pdf/1606.08415.pdf>`_.
+    """
+
+    def forward(self, x):
+        return torch.sigmoid(1.702 * x) * x
+
+
+class RightShift(nn.Module):
+    def __init__(self, embedding_dim):
+        super().__init__()
+        self.embedding_dim = embedding_dim
+        self.sos = nn.Parameter(
+            torch.FloatTensor(embedding_dim).normal_(std=0.02), requires_grad=True
+        )
+
+    def forward(self, x, decode_step=None):
+        if decode_step is not None and decode_step > 0:
+            return x
+
+        x_shape = list(x.shape)
+        x = x.flatten(start_dim=1, end_dim=-2)  # (batch, seq_len, embedding_dim)
+        sos = (
+            torch.ones(x_shape[0], 1, self.embedding_dim, dtype=torch.float32).to(
+                self.sos
+            )
+            * self.sos
+        )  # (batch, 1, embedding_dim)
+        sos = sos.type_as(x)
+        # Shift one unit to the right along dim ``seq_len``
+        x = torch.cat([sos, x[:, :-1, :]], axis=1)  # (batch, seq_len, embedding_dim)
+        x = x.view(*x_shape)
+        return x
+
+
+class MLPBlock(MLP):
+    """Transformer MLP block."""
+
+    def __init__(self, in_dim, out_dim, hidden_dims):
+        super().__init__(
+            in_dim, out_dim, hidden_dims, 0.0, activation=SiLU, normalization=None
+        )
+
+
+class TransformerLayer(nn.Module):
+    def __init__(
+        self,
+        shape,
+        embedding_dim,
+        num_attention_heads,
+        num_hidden_layers,
+        dropout,
+        attention_dropout,
+    ):
+        super().__init__()
+        self.pre_attention_norm = nn.LayerNorm(embedding_dim)
+        self.post_attention_dropout = nn.Dropout(dropout)
+        self.attention = MultiHeadAttention(
+            shape,
+            embedding_dim,
+            embedding_dim,
+            num_attention_heads,
+            num_hidden_layers,
+            causal=True,
+            attention_module=FullAttention(shape, True, attention_dropout),
+        )
+        self.pre_mlp_norm = nn.LayerNorm(embedding_dim)
+        self.post_mlp_dropout = nn.Dropout(dropout)
+        self.mlp_block = MLPBlock(embedding_dim, embedding_dim, [embedding_dim * 4])
+
+    def forward(self, x, decode_step, decode_idx):
+        h = self.pre_attention_norm(x)
+        if self.training:
+            h = checkpoint(self.attention, h, h, h, decode_step, decode_idx)
+        else:
+            h = self.attention(h, h, h, decode_step, decode_idx)
+        h = self.post_attention_dropout(h)
+        x = x + h
+
+        h = self.pre_mlp_norm(x)
+        if self.training:
+            h = checkpoint(self.mlp_block, h)
+        else:
+            h = self.mlp_block(h)
+        h = self.post_mlp_dropout(h)
+        x = x + h
+
+        return x
+
+
+class TransformerDecoder(nn.Module):
+    def __init__(
+        self,
+        position_embedding_input,
+        position_embedding_output,
```

Comment:
I'm assuming these are supposed to be of type `BroadcastedPositionEmbedding`? Are there use cases where a user will use some other position embedding? If so, there are assumptions that whatever the user feeds in will have `seq_len` and `embedding_dim` fields. If only `BroadcastedPositionEmbedding` is used, why not just instantiate it here so the user doesn't have to worry about it?